react/features/base/tracks/actions.web.ts

```diff
@@ -317,3 +292,225 @@ export function openAllowToggleCameraDialog(onAllow: Function, initiatorId: stri
+ *
+ * @param {JitsiLocalTrack} tracks - The tracks that have been created.
+ * @param {Function} dispatch - The redux dispatch function.
+ * @returns {void}
+ */
+export function setGUMPendingStateOnFailedTracks(tracks: Array<any>, dispatch: IStore['dispatch']) {
+    const tracksTypes = tracks.map(track => {
+        if (track.getVideoType() === VIDEO_TYPE.DESKTOP) {
+            return MEDIA_TYPE.SCREENSHARE;
+        }
+
+        return track.getType();
+    });
+    const nonPendingTracks = [ MEDIA_TYPE.AUDIO, MEDIA_TYPE.VIDEO ].filter(type => !tracksTypes.includes(type));
+
+    dispatch(gumPending(nonPendingTracks, IGUMPendingState.NONE));
+}
+
+/**
+ * Creates and adds to the conference the initial audio/video tracks.
+ *
+ * @param {Array<MediaType>} devices - Array with devices (audio/video) that will be used.
+ * @returns {Function}
+ */
+export function createAndAddInitialAVTracks(devices: Array<MediaType>) {
+    return (dispatch: IStore['dispatch']) => {
+        dispatch(gumPending(devices, IGUMPendingState.PENDING_UNMUTE));
+
+        return dispatch(createInitialAVTracks({ devices }))
+        .then(({ tracks, errors }) => {
+            setGUMPendingStateOnFailedTracks(tracks, dispatch);
+            dispatch(displayErrorsForCreateInitialLocalTracks(errors));
+
+            return Promise.allSettled(tracks.map((track: any) => {
+                const legacyConferenceObject = APP.conference;
+
+                if (track.isAudioTrack()) {
+                    return legacyConferenceObject.useAudioStream(track);
+                }
+                if (track.isVideoTrack()) {
+                    return legacyConferenceObject.useVideoStream(track);
+                }
+
+                return Promise.resolve();
+            }));
+        })
+        .finally(() => {
+            dispatch(gumPending(devices, IGUMPendingState.NONE));
+        });
+    };
+}
+
+/**
+ * Creates the initial audio/video tracks.
+ *
+ * @param {ICreateInitialTracksOptions} options - Options for creating the audio/video tracks.
+ * @returns {Function}
+ */
+export function createInitialAVTracks(options: ICreateInitialTracksOptions) {
+    return (dispatch: IStore['dispatch'], _getState: IStore['getState']) => {
+        const {
+            devices,
+            timeout,
+            firePermissionPromptIsShownEvent
+        } = options;
+        const audioOptions = {
+            devices: [ MEDIA_TYPE.AUDIO ],
+            timeout,
+            firePermissionPromptIsShownEvent
+        };
+
+        dispatch(gumPending(devices, IGUMPendingState.PENDING_UNMUTE));
+
+        return createLocalTracksF(options).then(tracks => {
+            return {
+                errors: {} as IInitialTracksErrors,
+                tracks
+            };
+        })
+        .catch(async error => {
+            const errors = {} as IInitialTracksErrors;
+
+            if (error.name === JitsiTrackErrors.TIMEOUT && !browser.isElectron()) {
```

Comment:
I know you are just moving code around, but Electron on macOS does show the permissions prompt, so I think we should remove the Electron check.

Reply:
Let me test this and think about it. Maybe we can do a check specifically for macOS electron. I'm a little bit worried to just remove it because the whole timeout logic was actually added because of Electron. There were many reports of issues that turned out to be related to GUM staying pending forever. Lets keep it for another PR?